File: test_src/fortran/hierarchical_parallelism/atomic/real/target__teams__loop.f90

program target__teams__loop
    USE OMP_LIB

    implicit none
  
    INTEGER :: L = 5
    INTEGER :: i
    
    REAL :: COUNTER = 0

    
    

     
    
    !$OMP TARGET   MAP(TOFROM: COUNTER) 



    

    
    !$OMP TEAMS 



    

    
    !$OMP LOOP 


    DO i = 1 , L 


    

    

!$OMP ATOMIC UPDATE

counter = counter + 1.


 
     

    END DO

    !$OMP END LOOP
     

    !$OMP END TEAMS
     

    !$OMP END TARGET
    

    IF  ( ( ABS(COUNTER - L) ) > 10*EPSILON(COUNTER) ) THEN
        write(*,*)  'Expected L Got', COUNTER
        call exit(1)
    ENDIF

end program target__teams__loop

File: test_src/cpp/hierarchical_parallelism/atomic/float/target__teams__loop.cpp
#include <iostream>
#include <cstdlib>
bool almost_equal(float x, float gold, float tol) {
        return gold * (1-tol) <= x && x <= gold * (1 + tol);
}
void test_target__teams__loop(){
 const int L = 262144;
 float counter{};
#pragma omp target  map(tofrom:counter) 
{
#pragma omp teams 
{
#pragma omp loop 
    for (int i = 0 ; i < L ; i++ )
{
#pragma omp atomic update
counter += float { 1.0f };
    } 
    } 
    } 
if ( !almost_equal(counter,float { L }, 0.1)  ) {
    std::cerr << "Expected: " << L << " Got: " << counter << std::endl;
    std::exit(112);
}
}
int main()
{
    test_target__teams__loop();
}
